### PR TITLE
Add support for decompressing .7z files

### LINKED ISSUE #27231: Add 7z/7zip file extraction support to Bazel
### Description of the feature request:

Bazel can extract variouse file types - e.g. *.zip, *.tar.gz, etc. It would be very helpfull if there would be also 7zip support (e.g. *.7z) files - extracting 7z files allways needs workaround: repacking, python scripts, non-hermetic thing (7zip installed on host system)

Not sure what is the best way - native support - or putting all file extraction thing in own rules - e.g. rules_7zip, rules_tar, neverthless this is recurring thing. For instance, in the past I had this issue when trying to support Qt  or Vulkan (since deps come in 7z files). Its always annoying that zip, tar, etc. works out of the box, but 7z not. 

### Which category does this issue belong to?

_No response_

### What underlying problem are you trying to solve with this feature?

7zip extraction

### Which operating system are you running Bazel on?

macOS, Ubunutu 24, Windows 11

### What is the output of `bazel info release`?

8.4.2

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

_No response_

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?

```text

```

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

### PR DESCRIPTION
**Pull request by willstranton**
This implementation follows the implementation of ZipDecompressor with the exception that .7z files do not handle symbolic links or preserving file permissions.  The commit 9c98120f33579b72561e02826d9fccf222eccb3c where .ar/.deb support was added, was used as a reference for what additional files to change.

Closes #27231

### CODE REVIEW

File: src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+import java.util.HashMap;
+
+import static com.google.common.truth.Truth.assertThat;
+import static com.google.devtools.build.lib.bazel.repository.decompressor.TestArchiveDescriptor.INNER_FOLDER_NAME;
+import static com.google.devtools.build.lib.bazel.repository.decompressor.TestArchiveDescriptor.ROOT_FOLDER_NAME;
+
+/** Tests .7z decompression. */
+@RunWith(JUnit4.class)
+public class SevenZDecompressorTest {
```
Comment (fmeum) [inline src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: Could you also add a test case for Unicode filename handling? You can search for examples of tests containing `äöü`.

Comment (willstranton) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: ok, see the added test `testUnicodeFilename` which checks that `ünïcödëFïlë.txt` is extracted.

**Comment by fmeum** [comment]
Not sure why, but ci says that you need to run "bazel run //src/test/tools/bzlmod:update_default_lock_file".

**Comment by willstranton** [comment]
> Not sure why, but ci says that you need to run "bazel run //src/test/tools/bzlmod:update_default_lock_file".

Ok, I ran it, and it changed `src/test/tools/bzlmod/MODULE.bazel.lock` to update the`bzlTransitiveDigest` which I assume is because I made edits in `tools/build_defs/repo/http.bzl` to update the comment.

Thanks for pointing that out!

File: src/main/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressor.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      SevenZArchiveEntry entry,
+      Path destinationDirectory,
+      PathFragment strippedRelativePath)
+      throws IOException, InterruptedException {
+    if (strippedRelativePath.isAbsolute()) {
+      throw new IOException(
+          String.format(
+              "Failed to extract %s, 7-zipped paths cannot be absolute", strippedRelativePath));
+    }
+    Path outputPath = destinationDirectory.getRelative(strippedRelativePath);
```
Comment (meteorcloudy) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressor.java:89]: Is it possible for `strippedRelativePath` to contain `../` and `outputPath` will actually be out of `destinationDirectory`? I think this might already be possible for other decompressors, but it'll be best if we can prevent it for new ones.

Comment (willstranton) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressor.java:89]: When a `PathFragment` is created, it should normalize the path:

eg. `a/b/../c` -> `a/c`

I guess multiple uplevel references could potentially escape the destination directory:

eg. `a/b/../../../../root` -> `../../root`

I added a check to make sure there can't be any uplevel references (`..`) when we create the `outputPath`

Comment (meteorcloudy) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressor.java:89]: Awesome, thank you!

**Review by meteorcloudy — APPROVED** [review]

File: src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java (hunk 1 above)
Comment (meteorcloudy) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: This test is failing internally, I'm getting
```
another_folder/ünïcödëFïlë.txt (No such file or directory)
```

Any idea how to fix?

Comment (fmeum) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: Could you modify the test you run internally so that it prints the entries in that directory and possibly also their UTF-8 byte representations?

Comment (willstranton) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: > This test is failing internally.

It's hard for me to help here when I don't know how internal testing differs from external testing.  Your help would be appreciated here!

> Any idea how to fix?

I did find one thing weird when I was originally writing the test. I had to call

`StringEncoding.unicodeToInternal(UNICODE_FILENAME)` to get the right filename in the filesystem.

this changed `ünïcödëFïlë.txt -> Ã¼nÃ¯cÃ¶dÃ«FÃ¯lÃ«.txt`

I wonder if that has something to do with it? 

In the updated test, I now run `StringEncoding::internalToUnicode` when reading filenames from the filesystem.  And when trying to match/find the internal unicode filename, I run the opposite: `StringEncoding.unicodeToInternal(UNICODE_FILENAME)`

> Could you modify the test you run internally so that it prints the entries in that directory

I did some test refactoring - see the latest two commits.  Now the test does a directory listing and shows what's there and the filename we expected.  Hope that helps meteorcloudy@ with more info from the internal testing.

Comment (willstranton) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: I added another commit to change calls of `internalToUnicode`/`unicodeToInternal` -> `internalToPlatform`/`platformToInternal` instead, which seems like the better choice since it checks if the platform even needs to conversion.  

 ¯\_(ツ)_/¯

File: src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  /** Test that Unicode filenames are handled. **/
+  @Test
+  public void testUnicodeFilename() throws Exception {
+    Path outputDir = decompress(archiveDescriptor().createDescriptorBuilder().build());
+
+    Path path = outputDir.getRelative(ROOT_FOLDER_NAME).getRelative(INNER_FOLDER_NAME);
+    List<String> filenames =
+        path.readdir(Symlinks.NOFOLLOW).stream()
+            .map((Dirent::getName))
+            .map(StringEncoding::internalToPlatform)
```
Comment (fmeum) [inline src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java]:
```suggestion
            .map(StringEncoding::internalToUnicode)
```
since you are passing the result to the Truth assertion libraries, which expect regular Java Unicode strings

File: src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            .collect(Collectors.toList());
+    assertThat(filenames).contains(UNICODE_FILENAME);
+
+    Path unicodeFile = path.getRelative(StringEncoding.platformToInternal(UNICODE_FILENAME));
```
Comment (fmeum) [inline src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java]:
```suggestion
    Path unicodeFile = path.getRelative(StringEncoding.unicodeToInternal(UNICODE_FILENAME));
```

File: src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java (hunk 3 above)
Comment (willstranton) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java]: ok, reverted this change.

File: src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java (hunk 4 above)
Comment (willstranton) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java]: ok, reverted this change.

File: src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java (hunk 1 above)
Comment (willstranton) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: ignore the last comment - I reverted this change per fmeum's comment: 

> since you are passing the result to the Truth assertion libraries, which expect regular Java Unicode strings

Comment (meteorcloudy) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]:
```
SevenZDecompressorTest_testDecompressWithRenamedFiles/root_folder/another_folder/ünïcödëFïlë.txt (No such file or directory)
	at com.google.devtools.build.lib.unix.NativePosixFiles.utimensat(Native Method)
	at com.google.devtools.build.lib.unix.UnixFileSystem.setLastModifiedTime(UnixFileSystem.java:364)
	at com.google.devtools.build.lib.vfs.Path.setLastModifiedTime(Path.java:734)
	at com.google.devtools.build.lib.bazel.repository.decompressor.SevenZDecompressor.extract7zEntry(SevenZDecompressor.java:115)
	at com.google.devtools.build.lib.bazel.repository.decompressor.SevenZDecompressor.decompress(SevenZDecompressor.java:65)
	at com.google.devtools.build.lib.bazel.repository.decompressor.SevenZDecompressorTest.decompress(SevenZDecompressorTest.java:142)
	at com.google.devtools.build.lib.bazel.repository.decompressor.SevenZDecompressorTest.testDecompressWithRenamedFiles(SevenZDecompressorTest.java:88)
```

I think it already failed during decompressing, so the directory listing didn't work.

Comment (willstranton) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: Thanks!  I added a debug commit that adds extra information (directory listing) to the exception that occurs when calling `setLastModifiedTime`.  Could you take a look and rerun internally to see what it outputs?

---

I tried to see if there was an explanation for the inconsistent behavior.  Here's one (possible red herring) thing I noticed:

There's a difference in the path when:

1) creating the outputstream to copy bytes into 
and
2) setting the modified time

For 1): When the filesystem is `UnixFileSystem` a `FileOutputStream` is constructed from a `File` that gets the path with `StringEncoding.internalToPlatform(path.getPathString())`:

https://github.com/bazelbuild/bazel/blob/bc86898ed9780702358715342dea81637c13e11c/src/main/java/com/google/devtools/build/lib/unix/UnixFileSystem.java#L436-L438

For 2) setting the modified time constructs the path with `path.toString()`:

https://github.com/bazelbuild/bazel/blob/bc86898ed9780702358715342dea81637c13e11c/src/main/java/com/google/devtools/build/lib/unix/UnixFileSystem.java#L361-L368

In 1) `path.getPathString()` is the same as 2) `toString()`:

1) `path.getPathString()`:

https://github.com/bazelbuild/bazel/blob/bc86898ed9780702358715342dea81637c13e11c/src/main/java/com/google/devtools/build/lib/vfs/PathFragment.java#L175-L177

2) `toString()`:

https://github.com/bazelbuild/bazel/blob/bc86898ed9780702358715342dea81637c13e11c/src/main/java/com/google/devtools/build/lib/vfs/PathFragment.java#L488-L490

---

So the difference is that 1) calls `StringEncoding.internalToPlatform()` on the underlying `normalizedPath` which passes it to `File`.  While 2) just passes `normalizedPath` to the native call `utimensat()`.  When I run this on my mac.  1 is passed the file path string:

1) <omitted...>/root_folder/another_folder/ünïcödëFïlë.txt

while 2 is passed the filepath string:

2) <omitted...>/root_folder/another_folder/Ã¼nÃ¯cÃ¶dÃ«FÃ¯lÃ«.txt

However, this just works somehow on my mac computer? When I do a raw directory listing, it's in the garbled (internal?) format:

`Ã¼nÃ¯cÃ¶dÃ«FÃ¯lÃ«.txt`

So I guess 1) ends up writing to that garbled (internal?) filepath somehow and it just works?    Curious what you'll get when you run it again.

---

Let me layout the possible paths to move forward here (let me know if you have other proposals):

1) Continue this back and forth debugging where I make changes and you run it internally 
2) You or another Google colleague take this change up and debug what's happening internally
3) Drop the unicode test and file (since no other decompressor does this test) - I wonder if the other decompressor tests would also fail when trying unicode files.

I'm happy to continue with 1 for maybe one more iteration, but I think 3 is better to get this change in (though there's no hurry).  2 seems like a better choice if you or a colleague have the time, since I'm running blind here (can't debug internally) and don't know the bazel codebase well.

Comment (meteorcloudy) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: Thanks! I will give it another try, if we still cannot figure out, let's remove this test case or make it Bazel only.

Comment (meteorcloudy) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]:
```
java.lang.RuntimeException: Found 2 files in directory of entry (root_folder/another_folder/Ã¼nÃ¯cÃ¶dÃ«FÃ¯lÃ«.txt): 
	?n?c?d?F?l?.txt
	renamedFile
	at com.google.devtools.build.lib.bazel.repository.decompressor.SevenZDecompressor.extract7zEntry(SevenZDecompressor.java:136)
	at com.google.devtools.build.lib.bazel.repository.decompressor.SevenZDecompressor.decompress(SevenZDecompressor.java:70)
	at com.google.devtools.build.lib.bazel.repository.decompressor.SevenZDecompressorTest.decompress(SevenZDecompressorTest.java:140)
	at com.google.devtools.build.lib.bazel.repository.decompressor.SevenZDecompressorTest.testDecompressWithRenamedFiles(SevenZDecompressorTest.java:87)
```

Looks like there was indeed some encoding issue.

Comment (meteorcloudy) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: The internal commons-compress is 1.27.1, the one in bazel is 1.26.1, but I guess that shouldn't be the reason?

Comment (fmeum) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: @willstranton Your findings above are all completely valid, but this is actually a red herring: `NativePosixFiles` directly operates on `String`s in the "internal" encoding (raw bytes) whereas `java.io.File` uses the "platform" encoding (usually UTF-8, but may also be raw bytes depending on your OS and locale settings). That's why the methods have to be called with different encodings, but they do look correct to me.

3. would be okay for me. The existing decompression tests for Unicode functionality are shell integration tests, not Java tests.

Comment (willstranton) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: > let's remove this test case or make it Bazel only.

I ended up removing the unicode file test case.  Just curious - how does one make a test "Bazel only"?  

> The existing decompression tests for Unicode functionality are shell integration tests, not Java tests.

I couldn't find existing integration tests which tested decompression under `src/test/shell/integration`

I saw you added some to `bazel_workspaces_test.sh` in 

https://github.com/bazelbuild/bazel/commit/10169bbe66e818868ec37f7d853ab5a567cd2ced#diff-c8e054134e31f6f0307f7dccc79dc185f43d5ce865d9c895232e388cf6992ab9

But that file and those tests ended up getting deleted in 37e0d239514c49c59a81501346e130b0330e8383 when the workspace->bzlmod transition happened.

Comment (meteorcloudy) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: > how does one make a test "Bazel only"?

In Java tests, we can check https://cs.opensource.google/bazel/bazel/+/master:src/test/java/com/google/devtools/build/lib/testutil/TestConstants.java;l=31

But I realize this isn't easy since the archive contains the unicode file name.

Comment (meteorcloudy) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: Tests passing internally now

Comment (fmeum) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: > But that file and those tests ended up getting deleted in https://github.com/bazelbuild/bazel/commit/37e0d239514c49c59a81501346e130b0330e8383 when the workspace->bzlmod transition happened.

Nice find, will add them back

Comment (fmeum) [reply src/test/java/com/google/devtools/build/lib/bazel/repository/decompressor/SevenZDecompressorTest.java:21]: https://github.com/bazelbuild/bazel/pull/27388
